### PR TITLE
SNOW-1805851: Add scikit-learn interoperability tests.

### PR DESCRIPTION
**Pull request by sfc-gh-mvashishtha**
- Move the existing scikit-learn interoperability test into the `interoperability` test folder.
- Refactor the test file so that it groups test cases into scikit-learn method categories, like classification and regression.
- Add test cases so we have one test case for each category.
- Remove the scikit-learn version pin for the modin dev environment. Once we implemented the interchange protocol, we no longer needed that pin.
- Document the support for the methods that we've tested.
- Reorganize the interoperability documentation page so that it follows the rst guidelines for sections and subsections.

### CODE REVIEW

**Review by sfc-gh-jrose — APPROVED** [review]
Test changes look real clean 🎉

File: docs/source/modin/interoperability.rst [hunk 1]
```diff
@@ -56,3 +59,95 @@ Plotly.express module methods
...
+
+
+Dimensionality reduction
+------------------------
+
++--------------------------------------------+---------------------------------------------+---------------------------------+
+| Operation                                  | Interoperable with Snowpark pandas? (Y/N)   | Notes for current implementation|
++--------------------------------------------+---------------------------------------------+---------------------------------+
+| Getting the principal components of a      | Y                                           |                                 |
+| numerical dataset with ``PCA.fit()``       |                                             |                                 |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/interoperability.rst:129]:
```suggestion
| numerical dataset with ``PCA.fit()``.      |                                             |                                 |
```

File: docs/source/modin/interoperability.rst [hunk 2]
```diff
@@ -56,3 +59,95 @@ Plotly.express module methods
...
+
+Model selection
+------------------------
+
++--------------------------------------------+---------------------------------------------+-----------------------------------------------+
+| Operation                                  | Interoperable with Snowpark pandas? (Y/N)   | Notes for current implementation              |
++--------------------------------------------+---------------------------------------------+-----------------------------------------------+
+| Choosing parameters for a                  | Y                                           | ``RandomizedSearchCV`` causes Snowpark pandas |
+| ``LogisticRegression`` model with          |                                             | to issue many queries. We strongly recommend  |
+| ``RandomizedSearchCV.fit()``               |                                             | converting Snowpark pandas inputs to pandas   |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/interoperability.rst:141]:
```suggestion
| ``RandomizedSearchCV.fit()``.              |                                             | converting Snowpark pandas inputs to pandas   |
```

**Review by sfc-gh-helmeleegy — APPROVED** [review]
Looks great. Thanks! Just left a few nits.

File: setup.py [hunk 3]
```diff
@@ -199,7 +199,7 @@ def run(self):
             *DEVELOPMENT_REQUIREMENTS,
             "scipy",  # Snowpark pandas 3rd party library testing
             "statsmodels",  # Snowpark pandas 3rd party library testing
-            "scikit-learn==1.5.2",  # Snowpark pandas scikit-learn tests
+            "scikit-learn",
```
Comment (sfc-gh-lmukhopadhyay) [inline setup.py:202]:
```suggestion
            "scikit-learn", # Snowpark pandas 3rd party library testing
```

File: tests/integ/modin/interoperability/scikit-learn/test_scikit_learn.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+4) Wrap the test with an empty sql_count_checker() decorator to see how many
+   queries and joins it requires. If it it requires a very large number of
+   queries, see if you can simplify the test case so that it causes fewer
+   queries, so that the test finishes quickly. If you can't reduce the number of
+   queries to a reasonable level, you should pass the SQL count checker the
+   `no_check=True` parameter because the number of queries is likely to vary
+   across scikit-learn and Snowpark pandas versions, and we don't gain much
+   insight by adjusting the query count every time it changes.
+5) Add a row describing interoperability with the new method in the
+   [documentation](docs/source/modin/interoperability.rst)
```
Comment (sfc-gh-lmukhopadhyay) [inline tests/integ/modin/interoperability/scikit-learn/test_scikit_learn.py:71]: Maybe you could also add info regarding best practices when the test result is deterministic vs non-deterministic? For ex, when to use `assert_numpy_results_valid` or another eval function?

**Review by sfc-gh-lmukhopadhyay — COMMENTED** [review]
Thanks, this looks great! Just left a couple of comments

File: docs/source/modin/interoperability.rst [hunk 5]
```diff
@@ -56,3 +59,95 @@ Plotly.express module methods
...
+| results with the ``predict()`` method.     |                                             |                                 |
++--------------------------------------------+---------------------------------------------+---------------------------------+
+
+Clustering
+----------
+
++--------------------------------------------+---------------------------------------------+---------------------------------+
+| Clustering method                          | Interoperable with Snowpark pandas? (Y/N)   | Notes for current implementation|
++--------------------------------------------+---------------------------------------------+---------------------------------+
+| ``KMeans.fit()``                           | Y                                           |                                 |
```
[the remark this answers is not in the document]
Comment (sfc-gh-mvashishtha) [reply docs/source/modin/interoperability.rst:118]: I considered that, but I decided to name the column here "Clustering Method", so I think it's clear to just put the method names in this column.

File: setup.py (hunk 3 above)
Comment (sfc-gh-mvashishtha) [reply setup.py]: Accepted. GitHub is being weird about accepting suggestions.

**Review by sfc-gh-lmukhopadhyay — APPROVED** [review]

**Pull request merged**